relation_engine/taxa/rdp/parsers.py

```diff
@@ -0,0 +1,159 @@
+            names, lineage = line.split('\t')
+            locus, definition = names.split(' ', 1)
+            lineage, unclassified = _get_lineage(lineage)
+            if not lineage:  # it's an outgroup
+                continue
+            for l in lineage:
+                l_id = _taxon_to_id(l)
+                if l_id not in seen_taxa:
+                    yield {
+                        'id': l_id.replace('/', '_'),
+                        'rank': l['rank'],
+                        'name': l['name'],
+                        'unclassified': False,
+                        'molecule': None,
+                        _INCERTAE_SEDIS: l[_INCERTAE_SEDIS]
+                    }
+                seen_taxa.add(l_id)
+            yield {
+                'id': locus[1:].strip(),  # remove '>'
+                'rank': 'sequence_example',
+                'name': definition.strip(),
+                'unclassified': unclassified,
+                'molecule': molecule,
+                _INCERTAE_SEDIS: None
+            }
+
+class RDPEdgeProvider:
+    """
+    RDPNodeProvider is an iterable that returns a new RDP taxonomy edge as a dict with each
+    iteration.
+    """
+
+    def __init__(self, rdp_taxonomy_file_handles):
+        """
+        Create the edge provider.
+
+        rdp_taxonomy_file_handles - a list of open handles for the RDP taxonomy files to process.
+        """
+        self._fh = rdp_taxonomy_file_handles
+
+    def __iter__(self):
+        seen_taxa = set()  # not including leaves
+        for fh in self._fh:
+            for line in fh:
+                if not line.startswith('>'):
+                    continue
+                names, lineage = line.split('\t')
+                locus, _ = names.split(' ', 1)
+                locus = locus[1:].strip()  # remove '>'
+                lineage, _ = _get_lineage(lineage)
+                if not lineage:  # it's an outgroup
+                    continue
+                for i in range(len(lineage) - 1):
+                    parent_id = _taxon_to_id(lineage[i]).replace('/', '_')
+                    child_id = _taxon_to_id(lineage[i + 1]).replace('/', '_')
+                    if child_id not in seen_taxa:
+                        yield {
+                            'id': child_id,  # one edge per child
+                            'from': child_id,
+                            'to': parent_id
+                        }
+                        seen_taxa.add(child_id)
+                parent_id = _taxon_to_id(lineage[-1]).replace('/', '_')
+                yield {
+                    'id': locus,  # one edge per child
+                    'from': locus,
+                    'to': parent_id
+                }
+
+# returns None in the first argument if the lineage indicates an outgroup.
+# second argument indicates if the sequence is unclassfied below the provided lineage
+def _get_lineage(linstr):
+    lin = linstr.replace('Lineage=', '')
+    l = lin.split(';')
```

Comment:
flake8 doesn't like this: https://lintlyci.github.io/Flake8Rules/rules/E741.html

Reply:
Fixed